### ISSUE TITLE
incorrect prediction from torchscript model converted from xgboost classifier trained with multi-label dataset 

### ISSUE THREAD

**Issue by louis-huang** [microsoft/hummingbird#719]
Hi, I tried to convert a xgboost model trained with multi-label data, but the prediction is not correct. I have a notebook [here](https://colab.research.google.com/drive/1HNPeQ6DinrT47aM4oAQ5rMMhwb35rFH4?usp=sharing).

I'm actually not sure if multi-label is supported in hummingbird, please help to verify this. Thank you!

**Comment by interesaaat** [comment]
Hi @louis-huang! Hummingbird supports multiclass for xgboost. You can check a test [here](https://github.com/microsoft/hummingbird/blob/7f4db051ccf6bca603824554cc95235926ab9929/tests/test_xgboost_converter.py#L79C5-L79C5). 

Hummingbird does not support categorical values for tree models. Can it be that the dataset has categorical features?

**Comment by louis-huang** [comment]
Hi @interesaaat , I want to try multi-label not multi-class model. This is an example from xgb: https://xgboost.readthedocs.io/en/stable/tutorials/multioutput.html

**Comment by gorkemozkaya** [comment]
@louis-huang The tree representations for a multi-class model vs. a multi-label classification model are essentially the same, with a separate tree ensemble for each class. So the Hummingbird for multi-class should work in theory. Some preprocessing may be necessary. 
```Internally, XGBoost builds one model for each target similar to sklearn meta estimators, with the added benefit of reusing data and other integrated features like SHAP. .```

**Comment by interesaaat** [comment]
Thanks @gorkemozkaya for chiming in. We already have some infra for multi-label for [sklearn multioutput regression](https://github.com/microsoft/hummingbird/blob/7f4db051ccf6bca603824554cc95235926ab9929/tests/test_sklearn_multioutput_regression.py#L24C5-L24C5). Shouldn't be too hard I feel to make it work for xgboost as well. Contributions welcome! 😄

**Comment by gorkemozkaya** [comment]
@interesaaat @louis-huang I think the problem is that Hummingbird is normalizing (i.e. applying softmax to) each row output so that the probabilities will add up to 1. If we could remove that normalization it would support multi-label. For now, I propose a workaround, by separating the n-way multi-label classifier into n separate binary classifiers, and the output probabilities match: [notebook link](https://github.com/gorkemozkaya/Data-Science-Notes/blob/master/reproducing_bugs/XGBoost_multi_label_classification_workaround_for_hummingbird_.ipynb)

**Comment by interesaaat** [comment]
We support [different post transformations](https://github.com/microsoft/hummingbird/blob/166f11efee12860870b74857831d8462d04b8227/hummingbird/ml/operator_converters/_gbdt_commons.py#L148). So maybe by setting `POST_TRANSFORM` to `None` it will remove the normalization step.

**Comment by gorkemozkaya** [comment]
Thanks @interesaaat, this was helpful! I verified that the outputs match if we override the default post-transform by passing the  `extra_config = {'post_transform': 'LOGISTIC'}`   argument, and then take the last `n_classes` columns of the output. 

By default it is using the `SOFTMAX` post_transform, which is not the right transform for multi-label. The library needs to decide the transform based on the `objective` attribute of the XGBoost classification model. I.e it should be changed such that `multi:softprob` maps to `SOFTMAX`, whereas `binary:logistic` maps to `LOGISTIC`. But it needs a slightly different version of `LOGISTIC` that does not double the number of output columns. For now, we can just take the last `n_classes` columns. 

<img width="1020" alt="image" src="https://github.com/microsoft/hummingbird/assets/6454229/38382a91-456d-4033-b1f6-c63ce584eac0">

**Comment by louis-huang** [comment]
Thanks for pointing out the post_transform @interesaaat This helps us a lot! Thank you @gorkemozkaya to verify this and providing suggestions!
